README.md
# Storefront Reference Architecture (SFRA)

This is a repository for the customized LyonsCG version of the Storefront Reference Architecture reference application.

Storefront Reference Architecture has a base cartridge (`app_storefront_base`) provided by Commerce Cloud that is never directly customized or edited. Instead, customization cartridges are layered on top of the base cartridge. This change is intended to allow for easier adoption of new features and bug fixes.
Storefront Reference Architecture supplies an [plugin_applepay](https://github.com/SalesforceCommerceCloud/plugin-applepay) plugin cartridge to demonstrate how to layer customizations for the reference application.

Please refer to our Confluence documentation for information and best practices regarding the new archtecture and how to override/extend server side and client side code.

SFRA Essentials Guide - https://lyonscg.atlassian.net/wiki/spaces/Intranet/pages/637372481/SFRA+Essentials+Guide+Draft

Client Side JavaScript Stack - https://lyonscg.atlassian.net/wiki/spaces/Intranet/pages/646985985/Client+Side+JavaScript+Stack+SFRA+Draft


The functional requirements for our LyonsCG version of SFRA are being maintained here - https://lyonscg.atlassian.net/wiki/spaces/SRV/pages/513802942/Functional+Requirements

## What's New

We keep our SFRA copy periodically updated with whatever version Salesforce considers "released". That version number is in our package.json file. You will see a "version" attribute near the top that indicates the Salesforce release we are in sync with, and an "lcgversion" which is our LyonsCG release number that will correspond with the list of tickets in the CHANGES.md file.

The CHANGES.md file offers an itemized list of internal ticket numbers that have been worked on for any particular release.

New features that we've added get a high level description and are maintained here - https://lyonscg.atlassian.net/wiki/spaces/Intranet/pages/568494465/SFRA+Finish+Grades+Features+Roadmap

## Getting Started

SFRA Setup Guide - https://lyonscg.atlassian.net/wiki/spaces/Intranet/pages/647171135/SFCC+Setup+Guide+SFRA+Under+Review

## Node version

The package.json file now specifies which version of node you should be running for installing our build tools.

---

## Quick Start

**A detailed setup guide is available! See [SFCC Setup Guide (SFRA)](https://lyonscg.atlassian.net/wiki/spaces/Intranet/pages/647171135?atlOrigin=eyJpIjoiYmJjYWNhNDdmYmUwNDE5Njg1YjE1YTg0NGIzYzJiNzAiLCJwIjoiYyJ9) on Confluence.**

**Complete documentation on SFRA npm scripts found in this Quick Start and the project is available in the [Build Tools README](./build_tools/README.md)**

### 1. Clone this repository.

```bash
git clone https://[YOUR-BITBUCKET-USERNAME]@bitbucket.org/lyonsconsultinggroup/reference-application-sfra.git
```

### 2. Install Project Dependencies

```bash
# move into root of project
cd reference-application-sfra

# install dependencies
npm install
```

### 3. Configure Upload Settings

Open [the dw.json example file](./dw.json.example)

You should see this JSON object:
```json
{
  "username": "myLogin",
  "password": "mySecretPassword",
  "clientId": "myClientIdApikey",
  "clientSecret": "myClientSecretPass",
  "hostname": ["dev01-na01-evaluation.demandware.net"],
  "activationHostname": ["dev01-na01-evaluation.demandware.net"],
  "code-version": "sfra",
  "dataBundle": "core"
}
```
Create a new dw.json file in the root directory by copying this example file.

The dw.json setup of the "username" and "password" depends on whether your environment has **Unified Authentication** enabled or not.

#### "Unified Authentication" Sandbox dw.json Settings

- `"username"`: Your Salesforce account manager username (email) i.e myUsername@lyonscg.com.
- `"password"`: Your Salesforce account manager password.

#### "No Unified Authentication" Sandbox dw.json Settings

- `"username"`: Your username to the specific sandbox instances i.e myUsername.
- `"password"`: The password set for the user on the specific sandbox instance.

#### Common dw.json Settings

- `"clientId"`: A valid Commerce Cloud API key which will be administered by the TA, TL, or lead developer. This is **required** for code-activation and site-import scripts; however, it is not required to upload code via VS Code's prophet debugger, Eclipse's digital server connection, or dwupload.
- `"clientSecret"`: A valid Commerce Cloud API secret which will be administered by the TA, TL, or lead developer. This is **required** for code-activation and site-import scripts; however, it is not required to upload code via VS Code's prophet debugger, Eclipse's digital server connection, or dwupload.
- `"hostname"`: An string hostname to the instance you are assigned. Code and data deployment scripts use this property 
    unless `deployHostname` has been set.
- `"deployHostname"`: An array of hostnames to the instances you want to deploy to. Code and data deployment scripts
    use this property first, if set, otherwise they will fall back to `hostname`. Can also be specified as a
    comma-delimited string.
- `"activationHostname"`: An array of hostnames to the instances you are assigned. Code version activation scripts use this property. Can also be specified as a comma-delimited string.
- `"code-version"`: The code version to deploy and activate code on.
- `"dataBundle"`: The data bundle(s) to deploy


### 4. OCAPI and WebDAV Settings

If you do not have a valid Commerce Cloud API key and secret (the `"clientId"` and `"clientSecret"` set in the dw.json), you can skip this step.

The code and deployment scripts depend on certain OCAPI and WebDav settings / permissions.

#### OCAPI

1. Navigate to Administration > Site Development > Open Commerce API Settings
2. Select the Data API (use the "Select type" drop-down)
3. Setup or confirm that a client object is setup with the following configuration:
```json
{
  "clients" : [ {
    "client_id" : "myClientIdApikey",
    "resources" : [ {
      "methods" : [ "get" ],
      "read_attributes" : "(**)",
      "write_attributes" : "(**)",
      "resource_id" : "/code_versions"
    }, {
      "methods" : [ "patch" ],
      "read_attributes" : "(**)",
      "write_attributes" : "(**)",
      "resource_id" : "/code_versions/*"
    }, {
      "methods" : [ "post" ],
      "read_attributes" : "(**)",
      "write_attributes" : "(**)",
      "resource_id" : "/jobs/*/executions"
    }, {
      "methods" : [ "get" ],
      "read_attributes" : "(**)",
      "write_attributes" : "(**)",
      "resource_id" : "/jobs/*/executions/*"
    } ]
  } ]
}
```

#### Webdav

1. Navigate to Adminstration > Organization > WebDAV Client Permissions
2. Setup of confirm that a client object is setup with the following configuration: 

```json
{
  "clients" : [ {
    "client_id" : "myClientIdApikey",
    "permissions" : [ {
      "path" : "/impex",
      "operations" : [ "read_write" ]
    }, {
      "path" : "/cartridges",
      "operations" : [ "read_write" ]
    } ]
  } ]
}

```

### 5. Sites and Cartridge Path Configuration

Open [the package.json file](./package.json)

Verify that the `"sites"` array property and the cartridge array per site object is properly configured. OOTB this property is configured correctly.  
If you're setting up a new project implementation see the [Build Tools README's](./build_tools/README.md) **Sites and Cartridge Path Configuration** section.


### 6. Compile Front-End Assets

```bash
# Compile fonts
npm run compile:fonts

# Compile client-side Js files
npm run compile:js

# Compile client-side Scss files
npm run compile:scss

# Compile the SVG sprite file
npm run compile:svg
```

You can verify the compilation of the Js and Css files by navigating to `cartridges/site_siteid/cartridge/static/default`

#### Pro front-end development tip: use the watcher!

```bash
# Watch for and compile any changes made to client-side JS or Sass files
npm run watch
```

### 7. Upload Code

It is highly recommended that you use your IDE tools to handle the cartridge uploading. VSCode's Prophet Debugger and the Eclipse server connection are going to be way more efficient at uploading code to your sandbox. The various upload commands included with our build tools are intended for a Jenkins server to be able to deploy code.

#### VS Code + Prophet Debugger Extension

Prophet Debugger uses your dw.json upload code (you do **not** need a clientId and/or clientSecret configured). It will watch for any file changes and upload automatically.

[Install Prophet Debugger](https://marketplace.visualstudio.com/items?itemName=SqrTT.prophet)

Additional documentation and setup for using VS Code as your IDE is available. See [Visual Studio Code - SFCC](https://lyonscg.atlassian.net/wiki/spaces/Intranet/pages/206670879/Visual+Studio+Code+-+SFCC?atlOrigin=eyJpIjoiNDU0N2U5YWJkMTc0NDkwOThhNmE2MzJmOWE0N2YwMWUiLCJwIjoiYyJ9) on confluence.

#### Eclipse Digital Server Connection

You can use Eclipse's digital server connection to upload your code. See the official [Salesforce B2C Docs](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/SiteDevelopment/UploadCartridges.html?resultof=%22%64%69%67%69%74%61%6c%22%20%22%64%69%67%69%74%22%20%22%73%65%72%76%65%72%22%20) on uploading cartridges.

#### LYONSCG deployCartridges script

If you have a valid Commerce Cloud API key and secret (the `"clientId"` and `"clientSecret"` set in the dw.json), you can run:

```bash
# Upload all cartridges and activate the code version that is specified in the dw.json
npm run deployCartridges
```

#### dwupload

The dwupload cli may be used to upload your code. See [dwupload on npm](https://www.npmjs.com/package/dwupload) for documentation.


### 8. Deploy Data

If you have a valid Commerce Cloud API key and secret (the `"clientId"` and `"clientSecret"` set in the dw.json), you can run:

```bash
# Deploy the data bundle that is specified in the dw.json
npm run deployData

```

Visit [the LYONSCG SFRA Data repository](https://bitbucket.org/lyonsconsultinggroup/sfra-storefront-data/) if you need a full set of static content (images), as that was omitted from this repository.

---

## NPM Scripts

**See the [Build Tools README](./build_tools/README.md)** for complete documentation on all the npm scripts available to aid in your development.

## Git Hooks

This application ships with a set if git hooks used to ensure standards (code and otherwise) are followed. Please see [husky](https://github.com/typicode/husky) for more info.

In the case of an **_emergency_**, you can bypass a git hook by passing the `--no-verify` flag. For example:

`git commit --no-verify`

## ESLint

ESLint is the linter of choice for JavaScript. It can be configured in two primary ways:

1. Configuration Comments (should rarely be used)
2. Configuration Files (primary method for configuring ESLint)

The configuration file used by LYONSCG (and the SFCC community) are `.eslintrc.json` files (not YAML, etc).

Searching through the codebase you'll note a number of `.eslintrc.json` files. Configuration files located higher in the file tree are extended by files lower in the tree.

When adding a LINK Cartridge (or other 3rd party cartridge) to the codebase: if the cartridge does not pass linting out-of-the-box and does not include it's own `.eslintrc.json` configuration file, add a suitable one to the cartridge root. It should stay as close as possible to the rules defined by LYONSCG (see the configuration file located in the root of this repo).

See the [ESLint user-guide](https://eslint.org/docs/user-guide/configuring) for more info.

## Running integration tests
Integration tests are located in the `storefront-reference-architecture/test/integration` directory.

To run integration tests you can use the following command:

**Note:** Please note that short form of this command will try to locate URL of your sandbox by reading `dw.json` file in the root directory of your project. If you don't have `dw.json` file, integration tests will fail.
sample `dw.json` file (this file needs to be in the root of your project)
{
    "hostname": "devxx-sitegenesis-dw.demandware.net"
}

You can also supply URL of the sandbox on the command line:

```
npm run test:integration -- --baseUrl devxx-sitegenesis-dw.demandware.net
```


# [Contributing to SFRA](./CONTRIBUTING.md)

#Page Designer Components for Storefront Reference Architecture
See: [Page Designer Components](./page-designer-components.md)


CHANGES.md
## 1.4.0 2/11/2020

- LRA-395 Update LRA with Salesforce SFRA version 4.4.1
- LRA-390 Add wishlist heart icon add/remove toggle behavior
- LRA-296 Fix image stacking on PDP quickview modal
- LRA-343 Add Account Locked Out handling and email template
- LRA-352 Add id with value prensentationID to link for custom search refinements
- LRA-387 Fix the sorting rule box to show the rule actually being applied
- LRA-349 Fix product compare checkbox on refinement reload
- LRA-389 Fix Page Context injection for replaced routes
- LRA-386 Merge changes in Salesforce SFRA 4.4
- LRA-385 Add PictureFill Support for IE
- LRA-242 Add Shipping Surcharge to Order Summary

## 1.3.0  9/27/2019

- LRA-382 Removed all references to "mfra"
- LRA-381 Updated Wishlist plugin
- LRA-378 Merged code with Salesforce SFRA version 4.3.0
- LRA-311 Optional quantity selector +/- controls
- LRA-372 Header Banner fix
- LRA-231 Confirm Email should not be required if Email is unchanged
- LRA-303 PDP Image Zoom fix
- LRA-374 Sitemap fix
- IEI-386 SARMS bug fix for duplicate components in multi-site project
- LRA-235 Source Code redirect added
- LRA-337 Fix to prevent skipping through checkout
- LRA-350 Sub category rendering template fix
- LRA-334 Tooltip fix
- LRA-332 SVG Builder added
- LRA-334 Merged code with Salesforce SFRA version 4.1.0
- LRA-293 Wishlist fix for mobile users

## 1.2.0  5/15/2019

- LRA-291	Updated LRA with Salesforce Releases 4.0.0, 3.6.0, 3.5.0, 3.4.0 - https://github.com/SalesforceCommerceCloud/storefront-reference-architecture/releases
- LRA-285	Slick carousel update to 1.8.1
- LRA-331	Build Tools: Single dw.json at the project root now works for build tools and VSCode with adjusted properties 
- LRA-241	Build Tools: Clean untracked files
- LRA-213	Build Tools: README.md updates
- LRA-246	Make search icon clickable
- LRA-287	PDP category locic check for variants
- LRA-314	Custom build report publishing to Confluence to replace outdated plugin
- LRA-213	Build Tools: sfcc-ci command line tool integrated into build tools to use new OCAPI methods for code and data deploy
- LRA-200	Build Tools: UTF-8 encoding fix
- LRA-290	Navigation look and feel changes across all devices
- LRA-279	Prevent double click for place order
- LRA-196	Add the missing Search-ShowContent controller route
- LRA-180	Build Tools: Adjustments for SARMS and Sonar Qube

## 1.1.2  3/21/2019

- LRA-302	Silence ESlint warnings
- LRA-305	Checkout Card Validation bugfix
- LRA-304	Product Compare regression fix

## 1.1.1  2/26/2019

- LRA-240	Build Tools:  (Webpack Enhancements) ES6 Support
- LRA-220	Enhanced Navigation SFRA
- LRA-84	Wishlist search from login page
- LRA-254	Fixing Global styling of action butttons
- LRA-248	Fixed footer so column headers don't appear clickable
- LRA-231	Account Edit Profile Form Confirm Email Not Required If Email Unchanged
- LRA-203	Open Tracking consent modal on click in edit profile
- LRA-207	CVV validation added/fixed for saved cards
- LRA-241	Build Tools: Ability to clean cached files in build script
- LRA-265	Item numbers now will show in mobile view
- LRA-269	Fixed image issues with Product set variation images
- LRA-256	Safari - Double scroll in the mini cart
- LRA-255	Fixed to show logout link on mobile
- LRA-198	Manual product reccomendations now implemented
- LRA-268	Fix to show social icons on pdp mobile
- LRA-233	Account Registration password error message fix
- LRA-249	Footer blocked on PDP bug fixed
- LRA-263	Product tile swatch - ellipsis click and carry through swatch
- IEI-180	Build Tools: Fixed SARMS reporting
- LRA-238	Link to PDP from cart line items
- LRA-272	Continue shopping link on cart

## 1.1.0  2/04/2019

- LRA-150	Responsive code fixes.
- LRA-158	Build Tools: removed default code value for data bundles so value can be inherited from dw.json
- LRA-119	Content asset in checkout
- LRA-117	Category browse breadcrumb path added
- LRA-168	Update data deploy script to use non-cert host for data import function
- LRA-165	Build Tools: verbose flag
- LRA-154	Styling issues
- LRA-155	Styling issues
- LRA-96	added toggle for product tile promo callout message
- LRA-170	Build Tools: Add default data bundle options for data_test and config_test
- LRA-141	SFRA oneline styleguide v2 update
- LRA-164	Build Tools: json validator
- LRA-178	Build Tools: only runs tests on cartridges with tests
- LRA-144	Feature to swap product images via swatches on cat browse
- LRA-146	SEO attributes on store detail page
- LRA-147	Store details page
- LRA-148	Contact us page with Recaptcha
- LRA-134	My account header
- IEI-211	GTM cartridge conversion
- IEI-253	Cascading PageContext
- IEI-252	Default PageContext config
- LRA-186	Product tile path fix
- LRA-181	Build Tools: option to minimize css
- LRA-192	utils.js updates to enhance build performance
- LRA-187	Moved homepage email signup to content asset
- LRA-145	Implement Zoom for PDP image, quickview, cart edit
- LRA-182	Build Tools: Improvements to CSS compilation time
- LRA-185	Build Tools: Improvements to JS compilation time
- LRA-162	Build Tools: Updating and annotating default site data
- LRA-190	Build Tools: Incorporating Webpack watcher
- LRA-189	Create category Grid template with content slot
- LRA-202	Build Tools: bitbucket pipeline yml added to run eslint
- LRA-193	bugfix for address book error message
- LRA-163	Build Tools: Webpack enhancements / Build script refactoring
- LRA-197	Salesforce 3.3.0 release tag merged in
- LRA-184	Resource message fix
- LRA-177	Build Tools: git pre-commit hooks using husky for linting
- LRA-207	Adding additional form validations to match SG functionality
- LRA-209	Build Tools: Webpack Enhancements with linting
- LRA-215	Manual Recommendations
- LRA ???	Fix Add My Account addressId check fix for adding and editing addresses
- LRA-201	Wishlist and Product compare added into our base offering
- LRA-216	Bonus Product Selection fixes
- LRA-236	Fixed multiple address editing issues in checkout

build_tools/util/webdav.js
"use strict";

const cheerio = require('cheerio');
const fs = require('fs');
const path = require('path');
const Q = require('q');
const querystring = require('querystring');
const request = require('request-promise-native').defaults({ simple: false });
const url = require('url');

/**
 * @class
 * @classdesc Webdav is used to perform upload and delete functions on the instance server that is provided by the configuration stored in the gulp builder object (gb)
 */
function Webdav(options) {
    /**
     * Contains the user name for the HTTP request
     * @member {string} user
     */
    this.username = options.username;

    /**
     * Contains the password for the HTTP request
     * @member {string} password
     */
    this.password = options.password;

    /**
     * Contains the host name of the webdav server
     * @member {string} hostname
     */
    this.hostname = 'https://' + (options.certHostname || options.hostname);
    
    /**
     * Contains the host name of the webdav server to activate code on
     * @member {string} activationHostname
     */
    this.activationHostname = 'https://' + options.activationHostname;

    /**
     * Contains the code version to use on the webdav server
     * Making an edit because the dwupload module and sgmf build scripts are out of sync 
     * with regards to variable naming
     */
    if ('codeVersion' in options) {
        this.codeVersion = options.codeVersion;
    }
    else if ('code-version' in options) {
        this.codeVersion = options['code-version'];
    }
    /**
     * Contains the p12 file location
     */
    this.p12 = options.p12;

    /**
     * Contains the p12 passphrase
     */
    this.passphrase = options.passphrase;

    /**
     * Contains the upload path for the webdav call
     */
    this.uploadPath = options.uploadPath || '/on/demandware.servlet/webdav/Sites/Impex/src/instance/';

    /**
     * Contains the url path for the impex status
     */
    this.impexStatusPath = '/on/demandware.store/Sites-Site/default/ViewSiteImpex-Status';

    /**
     * Contains the url path for the impex dispatch
     */
    this.impexDispatchPath = '/on/demandware.store/Sites-Site/default/ViewSiteImpex-Dispatch';

    /**
     * Contains the url path for the code activation script
     */
    this.activationPath = '/on/demandware.store/Sites-Site/default/ViewCodeDeployment-Activate';

    /**
     * Contains the url path for the business manager login form
     */
    this.loginPath = '/on/demandware.store/Sites-Site/default/ViewApplication-ProcessLogin';

    /**
     * The maximum number of times an HTTP request can retry
     * @member {number} MAX_RETRY
     */
    this.MAX_RETRY = 1;
};

/**
 * @function
 * @desc Logs into the business manager using the form instead of basic authentication
 * @returns
 */
Webdav.prototype.formLogin = async function () {
    const method = 'POST',
        deferred = Q.defer();

    // Login form should not include cert.staging
    var hostname = this.hostname;
    if (this.hostname.indexOf('cert.') > -1) {
        hostname = hostname.replace('cert.', '');
    }

    // Build the options for the upload
    let options = this.buildHttpOptions({
        method: method,
        uploadPath: this.loginPath,
        hostname: hostname
    });
    options.form = {
        LoginForm_Login: this.username,
        LoginForm_Password: this.password,
        LoginForm_RegistrationDomain: 'Sites'
    };
    options.jar = true;
    options.ignoreErrors = true;
    options.followRedirect = true;

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 302 || res.statusCode === 200) {
                console.log('Login Successful');
                deferred.resolve();
                return deferred.promise;
            } else {
                deferred.reject(new Error('Failed logged into ' + this.hostname + '. (Response: ' + res.statusCode + ')'));
                return;
            }
        });
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @description Uploads the data contained in the filename to the given destination
 * @param {string} filename - Zip file on the drive that contains the cartridges to be deployed to the given version
 * @param {string} dest - Destination that the file will be uploaded to
 */
Webdav.prototype.uploadFile = async function (filename, dest = path.basename(filename)) {
    const method = 'PUT',
        deferred = Q.defer();

    // Build the options for the upload
    let options = this.buildHttpOptions({
        method: method,
        dest: dest
    });
    options.body = fs.createReadStream(filename);

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 201 || res.statusCode === 200) {
                deferred.resolve();
            } else if (res.statusCode === 204) {
                let error = new Error('Remote file exists! The release file was not removed from the server on the last deployment. Please check to make sure that the unzip-files task worked correctly.');
                error.code = res.statusCode;
                deferred.reject(error);
                return;
            } else if (res.statusCode === 401) {
                let error = new Error('Authentication failed. Check your credentials.');
                error.code = res.statusCode;
                deferred.reject(error);
                return;
            } else if (res.statusCode === 405) {
                let error = new Error('Remote server does not support webdav!');
                error.code = res.statusCode;
                deferred.reject(error);
                return;
            } else {
                let error = new Error('Unknown error occurred:', res.statusCode);
                error.code = res.statusCode;
                deferred.reject(error);
                return;
            }
        });

        console.log('Deploying zip to ' + this.hostname + this.uploadPath + dest);
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @description Unzips the given file name on the server
 * @param {string} filename - Zip file on the drive that contains the cartridges to be deployed to the given version
 * @param {string} dest - Destination that the file will be uploaded to
 */
Webdav.prototype.unzipFile = async function (filename) {
    const method = 'POST',
        deferred = Q.defer();

    // Build the options for the upload
    let options = this.buildHttpOptions({
        method: method,
        dest: filename
    });
    options.form = {
        method: 'UNZIP'
    };

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 201 || res.statusCode === 200) {
                deferred.resolve();
            } else if (res.statusCode === 404) {
                let error = new Error('Given file', filename, 'not found');
                error.code = res.statusCode;
                deferred.reject(error);
                return;
            } else {
                let error = new Error('Unknown error occurred:', res.statusCode);
                error.code = res.statusCode;
                deferred.reject(error);
                return;
            }
        });
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @desc Checks to see if an import is ready to run
 * @param dest
 * @returns
 */
Webdav.prototype.importReady = async function () {
    const IMPORT_STATUS_RUNNING = 'Running';
    const IMPORT_PROCESS_NAME = 'Site Import';
    const method = 'GET',
    deferred = Q.defer();

    // Should not include cert.staging
    var hostname = this.hostname;
    if (this.hostname.indexOf('cert.') > -1) {
        hostname = hostname.replace('cert.', '');
    }

    // Build the options for the delete
    let options = this.buildHttpOptions({
        method: method,
        uploadPath: this.impexStatusPath,
        hostname: hostname
    });
    options.jar = true;
    options.rejectUnauthorized = false;

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 200 || res.statusCode === 201) {
                const $ = cheerio.load(res.body);
                let importReady = true;

                $('.table_detail:contains(' + IMPORT_PROCESS_NAME + ')').each(function () {
                    if ($(this).parent().find(':contains(' + IMPORT_STATUS_RUNNING + ')').length > 0) {
                        importReady = false;
                        deferred.reject(new Error('Previous import still in-process. Need to wait/retry.'));
                        return false;
                    }
                });

                if (importReady) {
                    console.log('Import is ready to go!');
                    deferred.resolve();
                }
            } else if (res.statusCode === 401) {
                deferred.reject(new Error('Authentication Failed'));
                return;
            } else if (res.statusCode === 404) {
                console.warn('Page not found');
                deferred.resolve();
            } else if (res.statusCode === 405) {
                deferred.reject(new Error('Remote server does not support webdav!'));
                return;
            } else {
                deferred.reject(new Error('Unknown error occurred:', res.statusCode));
                return;
            }
        });

        console.log('\nChecking if import is ready...');
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @desc Run the import for each zip file
 * @param dest
 * @returns
 */
Webdav.prototype.importFile = async function (archiveFilename) {
    const method = 'POST',
    deferred = Q.defer();

    // Should not include cert.staging
    var hostname = this.hostname;
    if (this.hostname.indexOf('cert.') > -1) {
        hostname = hostname.replace('cert.', '');
    }

    // Build the options for the delete
    let options = this.buildHttpOptions({
        method: method,
        uploadPath: this.impexDispatchPath,
        hostname: hostname
    });
    options.form = {
        ImportFileName: archiveFilename,
        realmUse: 'false',
        import: 'OK'
    };
    options.jar = true;
    options.rejectUnauthorized = false;

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 200) {
                console.log('Successful import of ' + archiveFilename);
                deferred.resolve();
                return deferred.promise;
            } else {
                console.log('Failed import of ' + archiveFilename);
                deferred.reject(new Error('Unknown Error: ' + response.statusCode));
                return;
            }
        });

        console.log('\nImporting ' + archiveFilename + '...');
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @description Removes the given destination from the server
 * @param {string} dest - Destination that will be removed from the server
 */
Webdav.prototype.deleteDestination = async function (dest = '') {
    const method = 'DELETE',
        deferred = Q.defer();

    // Build the options for the delete
    const options = this.buildHttpOptions({
        method: method,
        dest: dest
    });

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 204) {
                deferred.resolve();
            } else if (res.statusCode === 401) {
                deferred.reject(new Error('Authentication Failed'));
                return;
            } else if (res.statusCode === 404) {
                console.warn('Remote version did not exist! You may need to set overwriteRelease to false in the deployment config. Proceeding with build.');
                deferred.resolve();
            } else if (res.statusCode === 405) {
                deferred.reject(new Error('Remote server does not support webdav!'));
                return;
            } else {
                deferred.reject(new Error('Unknown error occurred:', res.statusCode));
                return;
            }
        });

        console.log('\nRemoving', options.url.path, 'from the server.');
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @description Activates the given code version
 * @param {String} version - The code version to activate
 */
Webdav.prototype.activateCode = async function (version = this.codeVersion) {
    const method = 'POST',
    deferred = Q.defer();

    // Build the options for the delete
    let options = this.buildHttpOptions({
        method: method,
        uploadPath: this.activationPath,
        hostname: this.activationHostname
    });
    options.form = {
        CodeVersionID: version
    };
    options.jar = true;
    options.rejectUnauthorized = false;

    if (options) {
        const req = this.handleRequest(options, (res) => {
            if (res.statusCode === 200) {
                // look for the active version check image to ensure code is activated
                var $ = cheerio.load(res.body);
                var activeVersionName = $('.table_detail img[title="Active version."]').parent('.table_detail').next('.table_detail').find('a.table_detail_link2').text();
                if (activeVersionName === version) {
                    console.log('Code version activated on ' + this.hostname + '. Deployment successful!');
                } else {
                    console.error('Error: code version not activated on ' + this.hostname + '. Check CSRF settings');
                }
                deferred.resolve();
                return;
            } else {
                deferred.reject(new Error('Unknown Error: ' + response.statusCode));
                return;
            }
        });

        console.log('\nActivating code at ' + version + '...');
    } else {
        deferred.reject(new Error('HTTP Options not properly setup.'));
    }

    return deferred.promise;
};

/**
 * @function
 * @description uploads the data contained in filename to the given destination
 * @param {object} options - Options used to build the HTTP request
 */
Webdav.prototype.buildHttpOptions = function (options) {
    // Check to make sure a user and password exist before continuing
    if (typeof this.username !== 'string' || typeof this.password !== 'string') {
        console.error('Required properties missing: `user` and `pass`.');
        return false;
    }

    options.hostname = options.hostname || this.hostname;
    options.uploadPath = options.uploadPath || this.uploadPath;
    options.dest = options.dest || '';

    const destUrl = options.hostname + options.uploadPath + options.dest;

    var deconstructedDest = url.parse(destUrl),
        httpOptions = {
            method : options.method,
            url: deconstructedDest,
            auth : {
                user: this.username,
                pass: this.password
            },
            resolveWithFullResponse: true
        };

    // Add some extra attributes if the server is using two factor authentication
    if (typeof this.p12 !== 'undefined' && destUrl.indexOf('https://cert') === 0) {
        console.log("Two Factor Enabled Using File: " + this.p12);
        httpOptions.pfx = fs.readFileSync(this.p12);
        httpOptions.passphrase = this.passphrase;
        httpOptions.honorCipherOrder = true;
        httpOptions.rejectUnauthorized = false;
        httpOptions.securityOptions = 'SSL_OP_NO_SSLv3';
    }

    return httpOptions;
};

/**
 * @function
 * @description uploads the data contained in filename to the given destination
 * @param {Object} options - HTTP options to be used in the request
 * @param {requestCallback} callback - Callback function that will handle the HTTP request
 */
Webdav.prototype.handleRequest = async function (options, callback) {
    // Turn off unauthorized rejections for TLS connections
    process.env.NODE_TLS_REJECT_UNAUTHORIZED = 0;

    let res = null

    try {
        res = await request(options);
        callback(res);
    } catch (e) {
        console.error('problem with request: ' + e.message);

        if (typeof this.p12 !== 'undefined' && this.hostname.indexOf('-') > -1) {
            console.warn('WARNING: Hyphens detected in hostname. Two-factor authentication requires usage of the dot convention in hostname.');
        }

        // Retry the request if an error has occurred up to the MAX_RETRY limit (default of 1)
        if (this.MAX_RETRY--) {
            this.handleRequest(options, callback);
        }
    }

    return res !== null ? res.request : null;
};

module.exports = Webdav;
